x/gamm/keeper/migrate_test.go

```diff
@@ -0,0 +1,359 @@
+				{
+					BalancerPoolId: 2,
+					ClPoolId:       7,
+				},
+				{
+					BalancerPoolId: 1,
+					ClPoolId:       6,
+				},
+			},
+			isPreexistingRecordsSet: true,
+			expectErr:               true,
+		},
+		{
+			name: "Normal case with two records",
+			testingMigrationRecords: []types.BalancerToConcentratedPoolLink{
+				{
+					BalancerPoolId: 1,
+					ClPoolId:       6,
+				},
+				{
+					BalancerPoolId: 2,
+					ClPoolId:       8,
+				},
+			},
+			expectedResultingRecords: []types.BalancerToConcentratedPoolLink{
+				{
+					BalancerPoolId: 1,
+					ClPoolId:       6,
+				},
+				{
+					BalancerPoolId: 2,
+					ClPoolId:       8,
+				},
+				{
+					BalancerPoolId: 3,
+					ClPoolId:       7,
+				},
+			},
+			isPreexistingRecordsSet: true,
+			expectErr:               false,
+		},
+		{
+			name: "Normal case with two records no preexisting records",
+			testingMigrationRecords: []types.BalancerToConcentratedPoolLink{
+				{
+					BalancerPoolId: 1,
+					ClPoolId:       6,
+				},
+				{
+					BalancerPoolId: 2,
+					ClPoolId:       8,
+				},
+			},
+			expectedResultingRecords: []types.BalancerToConcentratedPoolLink{
+				{
+					BalancerPoolId: 1,
+					ClPoolId:       6,
+				},
+				{
+					BalancerPoolId: 2,
+					ClPoolId:       8,
+				},
+			},
+			isPreexistingRecordsSet: false,
+			expectErr:               false,
+		},
+		{
+			name: "Modify existing record, delete existing record, leave a record alone, add new record",
+			testingMigrationRecords: []types.BalancerToConcentratedPoolLink{
+				{
+					BalancerPoolId: 1,
+					ClPoolId:       6,
+				},
+				{
+					BalancerPoolId: 2,
+					ClPoolId:       0,
+				},
+				{
+					BalancerPoolId: 4,
+					ClPoolId:       8,
+				},
+			},
+			expectedResultingRecords: []types.BalancerToConcentratedPoolLink{
+				{
+					BalancerPoolId: 1,
+					ClPoolId:       6,
+				},
+				{
+					BalancerPoolId: 3,
+					ClPoolId:       7,
+				},
+				{
+					BalancerPoolId: 4,
+					ClPoolId:       8,
+				},
+			},
+			isPreexistingRecordsSet: true,
+			expectErr:               false,
+		},
+		{
+			name: "Try to set one of the BalancerPoolIds to a cl pool Id",
+			testingMigrationRecords: []types.BalancerToConcentratedPoolLink{
+				{
+					BalancerPoolId: 2,
+					ClPoolId:       4,
+				},
+				{
+					BalancerPoolId: 5,
+					ClPoolId:       6,
+				},
+			},
+			isPreexistingRecordsSet: true,
+			expectErr:               true,
+		},
+	}
+
+	for _, test := range tests {
+		suite.Run(test.name, func() {
+			suite.SetupTest()
+			keeper := suite.App.GAMMKeeper
+
+			// Our testing environment is as follows:
+			// Balancer pool IDs: 1, 2, 3, 4
+			// Concentrated pool IDs: 5, 6, 7, 8
+			suite.PrepareBalancerPool()
+			suite.PrepareBalancerPool()
+			suite.PrepareBalancerPool()
+			suite.PrepareBalancerPool()
+			suite.PrepareConcentratedPool()
+			suite.PrepareConcentratedPool()
+			suite.PrepareConcentratedPool()
+			suite.PrepareConcentratedPool()
```

Comment:
We might want to have a helpers for initializing N pools in `apptesting`

Reply:
https://github.com/osmosis-labs/osmosis/pull/4071/commits/aa5638125e51843265a7bac5765f886f9c3c035d